test/system/src/test/java/io/pravega/test/system/ReadWriteAndScaleWithFailoverTest.java

```diff
@@ -0,0 +1,466 @@
+            testResult.get();
+        }
+
+    private void performFailoverTest() throws InterruptedException {
+
+        log.info("Test with 2 controller, SSS instances running and without a failover scenario");
+        long currentWriteCount1 = TestSate.EVENT_WRITE_COUNT.get();
+        long currentReadCount1 = TestSate.EVENT_READ_COUNT.get();
+        log.info("Read count: {}, write count: {} without any failover", currentReadCount1, currentWriteCount1);
+
+        //check reads and writes after some random time
+        int sleepTime = new Random().nextInt(50000) + 3000;
+        log.info("Sleeping for {} ", sleepTime);
+        Thread.sleep(sleepTime);
+
+        long currentWriteCount2 = TestSate.EVENT_WRITE_COUNT.get();
+        long currentReadCount2 =  TestSate.EVENT_READ_COUNT.get();
+        log.info("Read count: {}, write count: {} without any failover after sleep before scaling", currentReadCount2, currentWriteCount2);
+        //ensure writes are happening
+        assertTrue(currentWriteCount2 > currentWriteCount1);
+        //ensure reads are happening
+        assertTrue(currentReadCount2 > currentReadCount1);
+
+        //Scale down SSS instances to 2
+        segmentStoreInstance.scaleService(2, true);
+        //zookeeper will take about 60 seconds to detect that the node has gone down
+        Thread.sleep(ZK_DEFAULT_SESSION_TIMEOUT);
+        log.info("Scaling down SSS instances from 3 to 2");
+
+        currentWriteCount1 = TestSate.EVENT_WRITE_COUNT.get();
+        currentReadCount1 = TestSate.EVENT_READ_COUNT.get();
+        log.info("Read count: {}, write count: {} after SSS failover after sleep", currentReadCount1, currentWriteCount1);
+        //ensure writes are happening
+        assertTrue(currentWriteCount1 > currentWriteCount2);
+        //ensure reads are happening
+        assertTrue(currentReadCount1 > currentReadCount2);
+
+        //Scale down controller instances to 2
+        controllerInstance.scaleService(2, true);
+        //zookeeper will take about 60 seconds to detect that the node has gone down
+        Thread.sleep(ZK_DEFAULT_SESSION_TIMEOUT);
+        log.info("Scaling down controller instances from 3 to 2");
+
+        currentWriteCount2 = TestSate.EVENT_WRITE_COUNT.get();
+        currentReadCount2 = TestSate.EVENT_READ_COUNT.get();
+        log.info("Read count: {}, write count: {} after controller failover after sleep", currentReadCount2, currentWriteCount2);
+        //ensure writes are happening
+        assertTrue(currentWriteCount2 > currentWriteCount1);
+        //ensure reads are happening
+        assertTrue(currentReadCount2 > currentReadCount1);
+
+        //Scale down SSS, controller to 1 instance each.
+        segmentStoreInstance.scaleService(1, true);
+        controllerInstance.scaleService(1, true);
+        //zookeeper will take about 60 seconds to detect that the node has gone down
+        Thread.sleep(new Random().nextInt(50000) + ZK_DEFAULT_SESSION_TIMEOUT);
```

Comment:
We are waiting for zk to detect the node has gone down. `ZK_DEFAULT_SESSION_TIMEOUT` = 30s declared above.